CODING.md

```diff
@@ -0,0 +1,59 @@
+# Coding guide
+
+This project is written in Go programming language and follows a community accepted coding styles and practices defined in https://golang.org/doc/effective_go.html and https://github.com/golang/go/wiki/CodeReviewComments. This document defines additional guidelines specific to the Bee project.
+
+Developers should keep their Go tooling up to date with the latest stable versions.
+
+## Static code analysis
+
+Linting is done by using https://github.com/golangci/golangci-lint with configuration defined in the root of the project and also `go vet`. Executing `make lint vet` should pass without any warnings or errors.
+
+## Testing
+
+It is preferred to use separate test package to write tests. This allows explicit testing on exported behaviour of the package, not the internal implementation. In need to export only to testing package, export_test.go file should be used. Testing internal implementations discouraged, but not forbidden. Executing `make test` should pass without any warnings or errors.
```

Comment:
minor corrections:
- `to use a separate test package` instead of `to use separate test package`
- `internal implementaitons is discouraged` rather than `internal implementations discouraged`
- `expot_test.go` rather than export_test.go

and a question: why are we discouraging internal testing? isn't it better to have both?

Reply:
Thanks. Internal testing makes changes to the internal implementation harder, while behaviour testing on exposed api does not have to be changed when the internals change or improve.

Reply:
thanks for the clarification. i'm probably missing something though, because this sounds to me like you're saying "do not unit test"

Reply:
Maybe this would mean that it should be written a bi better. It should say that the unit tests should treat behaviours that package is exposing as units.

Reply:
requesting additional opinions here then (if it's just me i will resolve)

Reply:
if you have an unexported function that provides some very specific business logic that should not be exposed outside the package then you can and should unit test it that it performs according to your expectations. 

for all other functionalities - you should test at the edges of the package, and determine that the package or the contract as a whole _behaves_ as expected. this is what is called behavior driven testing or behavior driven development. 

it is better that the tests we write assume a certain _behavior_ rather than implementation details, this way, it is easier to determine if a component is functioning correctly but also allows for a component to be replaced by another implementation without much hassle, since interface contracts are to be maintained (see `SOLID` and `Liskov substitution principle` in particular)

Reply:
I guess that the biggest misunderstanding is what we call a `unit`. In `bee` I know that @janos is pushing for an approach where we try ask much as we can to call the _exposed_ unit our _unit_, rather than some underlying implementation details

Reply:
Thanks @acud. That is exactly what I wanted to say.

@mortelli I have added one sentence that should provide a context of a unit, but not to get into reasons for that.

I am happy to hear suggestions how this can be explained more clearly, if needed.